store/src/java/com/zimbra/cs/mailbox/LocalMailboxCacheManager.java

```diff
@@ -0,0 +1,296 @@
+package com.zimbra.cs.mailbox;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.zimbra.common.localconfig.LC;
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.common.util.ZimbraLog;
+import com.zimbra.cs.db.DbMailbox;
+import com.zimbra.cs.db.DbPool;
+
+import java.lang.ref.SoftReference;
+import java.util.*;
+
+public class LocalMailboxCacheManager implements MailboxCacheManager {
+    /**
+     * Maps account IDs (<code>String</code>s) to mailbox IDs
+     * (<code>Integer</code>s).  <i>Every</i> mailbox in existence on the
+     * server appears in this mapping.
+     */
+    private Map<String, Integer> mailboxIds;
+    /**
+     * Maps mailbox IDs ({@link Integer}s) to either
+     * <ul>
+     * <li>a loaded {@link Mailbox}, or
+     * <li>a {@link SoftReference} to a loaded {@link Mailbox}, or
+     * <li>a {@link MaintenanceContext} for the mailbox.
+     * </ul>
+     * Mailboxes are faulted into memory as needed, but may drop from memory when the SoftReference expires due to
+     * memory pressure combined with a lack of outstanding references to the {@link Mailbox}.  Only one {@link Mailbox}
+     * per user is cached, and only that {@link Mailbox} can process user requests.
+     */
+    private MailboxMap cache;
+
+    public LocalMailboxCacheManager() {
+        DbPool.DbConnection conn = null;
+        synchronized (this) {
```

Comment:
I think you need to move this synchronization into `MailboxManager`. Just make sure that the `MailboxCacheManager` constructor is protected in a synchronized block.

Reply:
moved to `MailboxManager`